File: backend/src/main/java/com/bobcarrental/service/BookingService.java
package com.bobcarrental.service;

import com.bobcarrental.dto.request.BookingRequest;
import com.bobcarrental.dto.response.BookingResponse;
import com.bobcarrental.dto.response.BookingSummaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;
import java.util.List;

/**
 * Service interface for Booking operations
 * 
 * @author Bob Car Rental System
 * @version 1.0
 */
public interface BookingService {

    /**
     * Create a new booking
     * 
     * @param request Booking creation request
     * @return Created booking
     */
    BookingResponse createBooking(BookingRequest request);

    /**
     * Update an existing booking
     * 
     * @param id Booking ID
     * @param request Booking update request
     * @return Updated booking
     */
    BookingResponse updateBooking(Long id, BookingRequest request);

    /**
     * Get booking by ID
     * 
     * @param id Booking ID
     * @return Booking details
     */
    BookingResponse getBookingById(Long id);

    /**
     * Get booking by booking number
     * 
     * @param bookingNo Booking number
     * @return Booking details
     */
    BookingResponse getBookingByBookingNo(String bookingNo);

    /**
     * Get all bookings with pagination
     * 
     * @param pageable Pagination parameters
     * @return Page of bookings
     */
    Page<BookingSummaryResponse> getAllBookings(Pageable pageable);

    /**
     * Get bookings by client ID
     * 
     * @param clientId Client ID
     * @return List of bookings
     */
    List<BookingSummaryResponse> getBookingsByClient(Long clientId);

    /**
     * Get bookings by status
     * 
     * @param status Booking status
     * @return List of bookings
     */
    List<BookingSummaryResponse> getBookingsByStatus(String status);

    /**
     * Get bookings by date range
     * 
     * @param startDate Start date
     * @param endDate End date
     * @return List of bookings
     */
    List<BookingSummaryResponse> getBookingsByDateRange(LocalDate startDate, LocalDate endDate);

    /**
     * Get upcoming bookings (from today onwards)
     * 
     * @return List of upcoming bookings
     */
    List<BookingSummaryResponse> getUpcomingBookings();

    /**
     * Cancel a booking
     * 
     * @param id Booking ID
     * @return Cancelled booking
     */
    BookingResponse cancelBooking(Long id);

    /**
     * Confirm a booking
     * 
     * @param id Booking ID
     * @return Confirmed booking
     */
    BookingResponse confirmBooking(Long id);

    /**
     * Complete a booking
     * 
     * @param id Booking ID
     * @return Completed booking
     */
    BookingResponse completeBooking(Long id);

    /**
     * Delete a booking
     * 
     * @param id Booking ID
     */
    void deleteBooking(Long id);

    /**
     * Check if booking number exists
     * 
     * @param bookingNo Booking number
     * @return true if exists, false otherwise
     */
    boolean existsByBookingNo(String bookingNo);

    /**
     * Validate booking availability
     *
     * @param typeId Vehicle type ID
     * @param fromDate From date
     * @param toDate To date
     * @return true if available, false otherwise
     */
    boolean isVehicleAvailable(String typeId, LocalDate fromDate, LocalDate toDate);
}

// Made with Bob

File: frontend/src/app/core/services/booking.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { environment } from '../../../environments/environment';
import { BookingResponse, BookingRequest, BookingSummaryResponse, ApiResponse } from '../../models';

interface PageResponse<T> {
  content: T[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
  first: boolean;
  last: boolean;
  hasPrevious: boolean;
  hasNext: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class BookingService {
  private apiUrl = `${environment.apiUrl}/bookings`;

  constructor(private http: HttpClient) {}

  /**
   * Get paginated list of bookings
   */
  getBookings(page: number = 0, size: number = 20, sort: string = 'bookDate,desc'): Observable<BookingSummaryResponse[]> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', sort);

    return this.http.get<ApiResponse<PageResponse<BookingSummaryResponse>>>(this.apiUrl, { params }).pipe(
      map(response => response.data.content)
    );
  }

  /**
   * Get booking by ID
   */
  getBookingById(id: number): Observable<BookingResponse> {
    return this.http.get<ApiResponse<BookingResponse>>(`${this.apiUrl}/${id}`).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get booking by booking number
   */
  getBookingByNumber(bookNum: string): Observable<BookingResponse> {
    return this.http.get<ApiResponse<BookingResponse>>(`${this.apiUrl}/by-number/${bookNum}`).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get bookings by client ID
   */
  getBookingsByClient(clientId: number): Observable<BookingSummaryResponse[]> {
    return this.http.get<ApiResponse<BookingSummaryResponse[]>>(`${this.apiUrl}/by-client/${clientId}`).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get bookings by status
   */
  getBookingsByStatus(status: string): Observable<BookingSummaryResponse[]> {
    return this.http.get<ApiResponse<BookingSummaryResponse[]>>(`${this.apiUrl}/by-status/${status}`).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get bookings by date range
   */
  getBookingsByDateRange(startDate: string, endDate: string): Observable<BookingSummaryResponse[]> {
    const params = new HttpParams()
      .set('startDate', startDate)
      .set('endDate', endDate);

    return this.http.get<ApiResponse<BookingSummaryResponse[]>>(`${this.apiUrl}/by-date-range`, { params }).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get upcoming bookings
   */
  getUpcomingBookings(): Observable<BookingSummaryResponse[]> {
    return this.http.get<ApiResponse<BookingSummaryResponse[]>>(`${this.apiUrl}/upcoming`).pipe(
      map(response => response.data)
    );
  }

  /**
   * Create new booking
   */
  createBooking(request: BookingRequest): Observable<BookingResponse> {
    return this.http.post<ApiResponse<BookingResponse>>(this.apiUrl, request).pipe(
      map(response => response.data)
    );
  }

  /**
   * Update existing booking
   */
  updateBooking(id: number, request: BookingRequest): Observable<BookingResponse> {
    return this.http.put<ApiResponse<BookingResponse>>(`${this.apiUrl}/${id}`, request).pipe(
      map(response => response.data)
    );
  }

  /**
   * Cancel booking
   */
  cancelBooking(id: number): Observable<BookingResponse> {
    return this.http.patch<ApiResponse<BookingResponse>>(`${this.apiUrl}/${id}/cancel`, {}).pipe(
      map(response => response.data)
    );
  }

  /**
   * Confirm booking (ADMIN only)
   */
  confirmBooking(id: number): Observable<BookingResponse> {
    return this.http.patch<ApiResponse<BookingResponse>>(`${this.apiUrl}/${id}/confirm`, {}).pipe(
      map(response => response.data)
    );
  }

  /**
   * Complete booking (ADMIN only)
   */
  completeBooking(id: number): Observable<BookingResponse> {
    return this.http.patch<ApiResponse<BookingResponse>>(`${this.apiUrl}/${id}/complete`, {}).pipe(
      map(response => response.data)
    );
  }

  /**
   * Delete booking (ADMIN only)
   */
  deleteBooking(id: number): Observable<void> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`).pipe(
      map(response => response.data)
    );
  }

  /**
   * Validate booking number
   */
  validateBookingNumber(bookNum: string): Observable<boolean> {
    const params = new HttpParams().set('bookingNo', bookNum);
    return this.http.get<ApiResponse<boolean>>(`${this.apiUrl}/validate/booking-no`, { params }).pipe(
      map(response => response.data)
    );
  }

  /**
   * Check vehicle availability
   */
  checkVehicleAvailability(vehicleTypeId: string, fromDate: string, toDate: string): Observable<boolean> {
    const params = new HttpParams()
      .set('vehicleTypeId', vehicleTypeId)
      .set('fromDate', fromDate)
      .set('toDate', toDate);

    return this.http.get<ApiResponse<boolean>>(`${this.apiUrl}/check-availability`, { params }).pipe(
      map(response => response.data)
    );
  }
}

// Made with Bob